### ISSUE TITLE
Error: `resource_id` and `type` are not matched

### ISSUE THREAD

**Issue by alexinthesky** [Azure/terraform-provider-azapi#149]
can someone help me out?

I'm trying to retrieve the tables of an LA workspace with the following code. VS code extensions suggested this type in the data block but apply ends up with 
**Error: `resource_id` and `type` are not matched**

```
resource "azurerm_log_analytics_workspace" "la_workspace" {
  name                = var.la_name
  location            = var.la_location
  resource_group_name = var.la_rg_name
  sku                 = var.la_sku
  retention_in_days   = var.la_retention
  tags                = var.la_tags
}

data "azapi_resource" "tables" {
  resource_id = resource.azurerm_log_analytics_workspace.la_workspace.id
  type      =  "Microsoft.OperationalInsights/workspaces/tables@2021-12-01-preview"
  response_export_values = ["name"]
}
```

what am I doing wrong?

**Comment by ms-henglu** [comment]
Hi @alexinthesky ,

Thank you for opening this issue!

As the error says, the type should be `Microsoft.OperationalInsights/workspaces@2021-12-01-preview`. But it seems that you want to list all the tables under an LA workspace, however this data source is not designed for this user case, it can only be used to read **one** existing resource.

**Comment by ms-henglu** [comment]
Hi @alexinthesky , Yes, because you want to retrieve the tables' names within an LA workspace, so in above tf config, I guess you mean
```
```

**Comment by kaovd** [comment]
Just adding my two cents as I see this isn't closed yet, but may need to add in some pre-checks that point the user to azapi_operation, as you can end up in a state where a resource gets created and the instance gets committed to the state, but then you are left in a limbo of "`resource_id` and `type` are not matched" till you remove it from the state manually

**Comment by ms-henglu** [comment]
Hi @kaovd ,

>  a resource gets created and the instance gets committed to the state, but then you are left in a limbo of "resource_id and type are not matched" till you remove it from the state manually

Would you please share me more details about this case? Maybe open a new issue about it?

Thanks!

**Comment by ms-henglu** [comment]
Hi @kaovd ,

>  a resource gets created and the instance gets committed to the state, but then you are left in a limbo of "resource_id and type are not matched" till you remove it from the state manually

Would you please share me more details about it? Maybe open another issue?

Thanks!

**Comment by kaovd** [comment]
@ms-henglu Heres a sample configuration

```hcl
resource "azurerm_monitor_private_link_scope" "example" {
  name                = "example-ampls"
  resource_group_name = azurerm_resource_group.example.name
}

resource "azapi_resource" "azurerm_monitor_private_link_scoped_service" {
  name = "endpoint"
  parent_id = azurerm_monitor_private_link_scope.example.id
  type = "Microsoft.Insights/privatelinkscopes/scopedResources@2021-07-01-preview"

  body = jsonencode({
    properties = {
        linkedResourceId = jsondecode(data.azapi_resource.dce.output).id
    }
  })

  depends_on = [
    azurerm_monitor_private_link_scope.example
  ]

}
```

I am attempting to create an azure monitor resource scope connection from a AMPLS to a data collection endpoint - which is a PUT Operation although is technically a [resource](https://github.com/Azure/azure-sdk-for-go/blob/b386adc15c2e60e3812956f68d244b608330c479/services/preview/monitor/mgmt/2021-07-01-preview/insights/privatelinkscopedresources.go#L93) and can do standard restful args like GET, DELETE, etc. (linkedResourceId being a data collection endpoint - `microsoft.insights/datacollectionendpoints`, `parentid` in the format of `/subscriptions/****/resourceGroups/****/providers/Microsoft.Insights/privateLinkScopes/example-ampls` )

The apply will succeed, and the resource will get created 
![image](https://user-images.githubusercontent.com/29356754/184359121-0bd52454-51e8-4739-a616-6fabea8a001c.png)

However the apply will fail with the given error and any other terraform operations will fail until you pull it from the state

Understand this is a similar issue where it is expected it to be Insights/privateLinkScopes and not additionally scoped resources, although in this case it is just a nested resource and not multiple or just an operation.

**Comment by ms-henglu** [comment]
Hi @kaovd ,

Thank you so much for the example, I did some investigation, and found the `type`'s casing is incorrect, please check the following hcl. Please notice the casing of `privateLinkScopes`, also it's highly recommended to use [azapi vscode extension](https://marketplace.visualstudio.com/items?itemName=azapi-vscode.azapi) which can help you auto-complete the type and its properties.

```hcl
resource "azapi_resource" "endpoint" {
  name      = "endpoint"
  parent_id = azurerm_monitor_private_link_scope.test.id
  type      = "Microsoft.Insights/privateLinkScopes/scopedResources@2021-07-01-preview"

  body = jsonencode({
    properties = {
      linkedResourceId = azapi_resource.dce.id
    }
  })
  // properties.linkedResourceId is not returned in correct casing
  ignore_casing = true
}
```

This is also an azapi bug that it has inconsistent `type` validation, I'll fix it.

**Comment by donwlewis** [comment]
@ms-henglu I also ran into this when attempting to call to the `Microsoft.Cdn/profiles/afdendpoints/routes@2020-09-01`.  Setting `ignore_casing` seems to have no affect.  Is it potentially related to the bug you mentioned?  Thanks.

**Comment by ms-henglu** [comment]
Hi @donwlewis , Yes, please check: https://github.com/Azure/terraform-provider-azapi/issues/147#issuecomment-1176981994

And the document for azapi provider just go live, now you can browse all the available resource types and their properties from https://docs.microsoft.com/en-us/azure/templates/microsoft.cdn/profiles/afdendpoints/routes?pivots=deployment-language-terraform

**Comment by ms-henglu** [comment]
Hi @alexinthesky,

The [azapi_resource_action](https://registry.terraform.io/providers/Azure/azapi/latest/docs/resources/azapi_resource_action) has been released in azapi 0.5.0, would you please give it a try? And there's an example to list tables under an LA workspace.
https://github.com/Azure/terraform-provider-azapi/blob/main/examples/Microsoft.OperationalInsights/list_workspace_tables/main.tf
